### PR TITLE
fix(components): [date-picker] disabledDate is invalid when selecting year or month picker

### LINKED ISSUE #10604: [Component] [date-picker] DatePicker组件设置禁用时间，仍然可以通过顶部月份入口默认选中可选月份第一天！
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.45`
- Element Plus Version: `2.2.21`
- Browser / OS: `Chrome/107.0.0.0 / Macintosh; Intel Mac OS X 10_15_7`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2IGNsYXNzPVwiZGVtby1kYXRlLXBpY2tlclwiPlxuICAgIDxkaXYgY2xhc3M9XCJjb250YWluZXJcIj5cbiAgICAgICA8ZGl2IGNsYXNzPVwiYmxvY2tcIj5cbiAgICAgICAgPHNwYW4gY2xhc3M9XCJkZW1vbnN0cmF0aW9uXCI+ZGF5PC9zcGFuPlxuICAgICAgICA8ZWwtZGF0ZS1waWNrZXJcbiAgICAgICAgICB2LW1vZGVsPVwidmFsdWUxXCJcbiAgICAgICAgICB0eXBlPVwiZGF0ZVwiXG4gICAgICAgICAgZm9ybWF0PVwiTU0vREQvWVlZWVwiXG4gICAgICAgICAgOmRpc2FibGVkLWRhdGU9XCJzZGRcIlxuICAgICAgICAgIHBsYWNlaG9sZGVyPVwiUGljayBhIGRheVwiXG4gICAgICAgIC8+XG4gICAgICA8L2Rpdj5cbiAgICA8L2Rpdj5cbiAgPC9kaXY+XG48L3RlbXBsYXRlPlxuPHNjcmlwdCBsYW5nPVwidHNcIiBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgdmFsdWUxID0gcmVmKCcnKVxuY29uc3QgdmFsdWUyID0gcmVmKCcnKVxuY29uc3QgdmFsdWUzID0gcmVmKCcnKVxuY29uc3QgdmFsdWU0ID0gcmVmKCcnKVxuXG5jb25zdCBzZGQgPSAodGltZTogRGF0ZSkgPT4ge1xuICByZXR1cm4gdGltZS5nZXRUaW1lKCkgPCBEYXRlLm5vdygpXG59XG48L3NjcmlwdD5cbjxzdHlsZSBzY29wZWQ+XG4uZGVtby1kYXRlLXBpY2tlciB7XG4gIGRpc3BsYXk6IGZsZXg7XG4gIHdpZHRoOiAxMDAlO1xuICBwYWRkaW5nOiAwO1xuICBmbGV4LXdyYXA6IHdyYXA7XG59XG5cbi5kZW1vLWRhdGUtcGlja2VyIC5ibG9jayB7XG4gIHBhZGRpbmc6IDMwcHggMDtcbiAgdGV4dC1hbGlnbjogY2VudGVyO1xuICBib3JkZXItcmlnaHQ6IHNvbGlkIDFweCB2YXIoLS1lbC1ib3JkZXItY29sb3IpO1xuICBmbGV4OiAxO1xufVxuLmRlbW8tZGF0ZS1waWNrZXIgLmJsb2NrOmxhc3QtY2hpbGQge1xuICBib3JkZXItcmlnaHQ6IG5vbmU7XG59XG5cbi5kZW1vLWRhdGUtcGlja2VyIC5jb250YWluZXIge1xuICBmbGV4OiAxO1xuICBib3JkZXItcmlnaHQ6IHNvbGlkIDFweCB2YXIoLS1lbC1ib3JkZXItY29sb3IpO1xufVxuLmRlbW8tZGF0ZS1waWNrZXIgLmNvbnRhaW5lciAuYmxvY2sge1xuICBib3JkZXItcmlnaHQ6IG5vbmU7XG59XG4uZGVtby1kYXRlLXBpY2tlciAuY29udGFpbmVyIC5ibG9jazpsYXN0LWNoaWxkIHtcbiAgYm9yZGVyLXRvcDogc29saWQgMXB4IHZhcigtLWVsLWJvcmRlci1jb2xvcik7XG59XG4uZGVtby1kYXRlLXBpY2tlciAuY29udGFpbmVyOmxhc3QtY2hpbGQge1xuICBib3JkZXItcmlnaHQ6IG5vbmU7XG59XG5cbi5kZW1vLWRhdGUtcGlja2VyIC5kZW1vbnN0cmF0aW9uIHtcbiAgZGlzcGxheTogYmxvY2s7XG4gIGNvbG9yOiB2YXIoLS1lbC10ZXh0LWNvbG9yLXNlY29uZGFyeSk7XG4gIGZvbnQtc2l6ZTogMTRweDtcbiAgbWFyZ2luLWJvdHRvbTogMjBweDtcbn1cbjwvc3R5bGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

1、点击日期选择顶部入口【月份】
2、选择可选月份，默认会显示月份第一天（此日期为禁用日期）

### What is Expected?

选择对应月份后，显示可选日期的第一天

### What is actually happening?

显示月份第一天（1号）

### Additional comments

_(empty)_

### LINKED ISSUE #11413: [Component] [date-picker] date-picker  type="date" 选择年份或月份时会立刻触发双向绑定的日期数据更新。导致禁用的日期被选中
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.47`
- Element Plus Version: `2.2.29`
- Browser / OS: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgdmFsdWUgPSByZWYoJzIwMjMtMDItMDQnKVxuY29uc3QgZGF5VGltZSA9IG5ldyBEYXRlKCcyMDIzLTAyLTAzJykuZ2V0VGltZSgpXG5jb25zdCBkaXNhYmxlZERhdGUgPSAodGltZTogRGF0ZSkgPT4ge1xuICByZXR1cm4gdGltZS5nZXRUaW1lKCkgPCBkYXlUaW1lXG59XG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8ZWwtZGF0ZS1waWNrZXJcbiAgICB2LW1vZGVsPVwidmFsdWVcIlxuICAgIHR5cGU9XCJkYXRlXCJcbiAgICBwbGFjZWhvbGRlcj1cIlBpY2sgYSBkYXlcIlxuICAgIDpkaXNhYmxlZC1kYXRlPVwiZGlzYWJsZWREYXRlXCJcbiAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

el-date-picker禁用2023/02/03之前的日期，当前的绑定的日期是2024/01/01，点击年份year-table弹框，选择2023年。绑定的时间更新并选中了禁用的日期

### What is Expected?

直到选中了具体的某一天时，绑定日期数据才进行更新

### What is actually happening?

选择年份或月份时，数据被立即更新

### PR DESCRIPTION
**Pull request by xingyixiang**
### description:
此次修改不改变现有的交互。当通过年份面板或月份面板获取的日期是被禁用时，会找到当前年份、月份中最早的可用时间。因为复用了datesInMonth方法，将其抽离到utils.ts文件中。并对此修改特性增加了测试用例。

This modification does not change the existing interaction. When the date obtained through the year or month panel is disabled, the earliest available time in the current year or month will be found. Because the datesInMonth method was reused, it was extracted into the utils.ts file. And added test cases to describe this feature.

Please make sure these boxes are checked before submitting your PR, thank you!
fixes: #11413 
fixes: #10604

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by xingyixiang** [comment]
[Example](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6IjxzY3JpcHQgc2V0dXAgbGFuZz1cInRzXCI+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5cbmNvbnN0IHZhbHVlID0gcmVmKCcyMDIzLTAyLTA0JylcbmNvbnN0IHZhbHVlMSA9IHJlZignMjAyNC0wMi0wMScpXG5jb25zdCBkYXlUaW1lID0gbmV3IERhdGUoJzIwMjMtMDItMDMnKS5nZXRUaW1lKClcbmNvbnN0IGRpc2FibGVkRGF0ZSA9ICh0aW1lOiBEYXRlKSA9PiB7XG4gIHJldHVybiB0aW1lLmdldFRpbWUoKSA8IGRheVRpbWVcbn1cbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxwPnNlbGVjdCBtb250aCBwaWNrPC9wPlxuICA8ZWwtZGF0ZS1waWNrZXJcbiAgICB2LW1vZGVsPVwidmFsdWVcIlxuICAgIHR5cGU9XCJkYXRlXCJcbiAgICBwbGFjZWhvbGRlcj1cIlBpY2sgYSBkYXlcIlxuICAgIDpkaXNhYmxlZC1kYXRlPVwiZGlzYWJsZWREYXRlXCJcbiAgLz5cblxuICA8cD5zZWxlY3QgeWVhciBwaWNrPC9wPlxuICA8ZWwtZGF0ZS1waWNrZXJcbiAgICB2LW1vZGVsPVwidmFsdWUxXCJcbiAgICB0eXBlPVwiZGF0ZVwiXG4gICAgcGxhY2Vob2xkZXI9XCJQaWNrIGEgZGF5XCJcbiAgICA6ZGlzYWJsZWQtZGF0ZT1cImRpc2FibGVkRGF0ZVwiXG4gIC8+XG48L3RlbXBsYXRlPlxuIiwic3JjL1BsYXlncm91bmRNYWluLnZ1ZSI6IjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgQXBwIGZyb20gJy4vQXBwLnZ1ZSdcbmltcG9ydCB7IHNldHVwRWxlbWVudFBsdXMgfSBmcm9tICcuL2VsZW1lbnQtcGx1cy5qcydcbnNldHVwRWxlbWVudFBsdXMoKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPEFwcCAvPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vcHJldmlldy0xNTg0OC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcInVuc3VwcG9ydGVkXCJcbiAgfVxufSIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJzcmMvZWxlbWVudC1wbHVzLmpzIjoiaW1wb3J0IHsgZ2V0Q3VycmVudEluc3RhbmNlIH0gZnJvbSAndnVlJ1xuaW1wb3J0IEVsZW1lbnRQbHVzIGZyb20gJ2VsZW1lbnQtcGx1cydcblxubGV0IGluc3RhbGxlZCA9IGZhbHNlXG5hd2FpdCBsb2FkU3R5bGUoKVxuXG5leHBvcnQgZnVuY3Rpb24gc2V0dXBFbGVtZW50UGx1cygpIHtcbiAgaWYgKGluc3RhbGxlZCkgcmV0dXJuXG4gIGNvbnN0IGluc3RhbmNlID0gZ2V0Q3VycmVudEluc3RhbmNlKClcbiAgaW5zdGFuY2UuYXBwQ29udGV4dC5hcHAudXNlKEVsZW1lbnRQbHVzKVxuICBpbnN0YWxsZWQgPSB0cnVlXG59XG5cbmV4cG9ydCBmdW5jdGlvbiBsb2FkU3R5bGUoKSB7XG4gIGNvbnN0IHN0eWxlcyA9IFsnaHR0cHM6Ly9wcmV2aWV3LTE1ODQ4LWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguY3NzJywgJ2h0dHBzOi8vcHJldmlldy0xNTg0OC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmNzcyddLm1hcCgoc3R5bGUpID0+IHtcbiAgICByZXR1cm4gbmV3IFByb21pc2UoKHJlc29sdmUsIHJlamVjdCkgPT4ge1xuICAgICAgY29uc3QgbGluayA9IGRvY3VtZW50LmNyZWF0ZUVsZW1lbnQoJ2xpbmsnKVxuICAgICAgbGluay5yZWwgPSAnc3R5bGVzaGVldCdcbiAgICAgIGxpbmsuaHJlZiA9IHN0eWxlXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2xvYWQnLCByZXNvbHZlKVxuICAgICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdlcnJvcicsIHJlamVjdClcbiAgICAgIGRvY3VtZW50LmJvZHkuYXBwZW5kKGxpbmspXG4gICAgfSlcbiAgfSlcbiAgcmV0dXJuIFByb21pc2UuYWxsKHN0eWxlcylcbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy0xNTg0OC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmNzcyJ9fQ==)

**Comment by xingyixiang** [comment]
cc @kooriookami @btea @FrontEndDog

File: packages/components/date-picker/src/date-picker-com/panel-date-pick.vue [hunk 1]
```diff
@@ -405,8 +406,37 @@ const keyboardMode = computed<string>(() => {
...
+    return getValidDateOfMonth(month)
+  }
+  for (let i = 0; i < 12; i++) {
+    if (!datesInMonth(year, i, lang.value).every(disabledDate)) {
+      return getValidDateOfMonth(i)
+    }
+  }
+  return innerDate.value
+}
+
```
Comment (btea) [inline packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:437]: Perhaps these two methods can also be extracted into `utils`.

Comment (xingyixiang) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:437]: In this case, disabledDate, innerDate, and lang need to be parameters in the current function. I don't think it's very good.

Comment (btea) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:437]: Ok.

**Review by btea — APPROVED** [review]
LGTM. Thanks!

**Comment by xingyixiang** [comment]
@btea 有点问题：在某一个月中进行日期遍历时没有考虑到locale的影响，有些国家的月份天数是不一样的。我找时间修改一下
There is a bit of a problem that date traversal in a month does not take into account the effect of locale. Some countries have different number of days in a month. I'll find time to revise it.

**Comment by xingyixiang** [comment]
@btea The getValidDateOfMonth method does not consider the impact of locale. It's fixed now.

**Review by FrontEndDog — APPROVED** [review]
Great！ Thanks！

File: packages/components/date-picker/src/date-picker-com/panel-date-pick.vue [hunk 2]
```diff
@@ -416,17 +417,47 @@ const keyboardMode = computed<string>(() => {
...
 const handleMonthPick = async (
   month: number | MonthsPickerEmits,
   keepOpen?: boolean
 ) => {
   if (selectionMode.value === 'month') {
-    innerDate.value = innerDate.value.startOf('month').month(month as number)
+    innerDate.value = getValidDateOfMonth(month as number)
     emit(innerDate.value, false)
   } else if (selectionMode.value === 'months') {
     emit(month as MonthsPickerEmits, keepOpen ?? true)
```
Comment (tolking) [inline packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:458]: 我注意到这里没有对 months 进行处理，当 `type="months"` 时依然会出现类似问题。我觉得或许还应该对 months、years 进行处理。

Comment (xingyixiang) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:458]: type为months、years 我理解不会出现这个问题，这个只有当选择完年或月需要继续选择月和日的情况才会出现问题

Comment (tolking) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:458]: [Demo](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgdmFsdWUgPSByZWYoKVxuY29uc3QgdmFsdWUxID0gcmVmPERhdGU+KClcbmNvbnN0IHZhbHVlMiA9IHJlZjxEYXRlPigpXG5jb25zdCBkYXlUaW1lID0gbmV3IERhdGUoMjAyNCwgNywgNCkuZ2V0VGltZSgpXG5jb25zdCBkaXNhYmxlZERhdGUgPSAodGltZTogRGF0ZSkgPT4ge1xuICByZXR1cm4gdGltZS5nZXRUaW1lKCkgPCBkYXlUaW1lXG59XG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8ZWwtZGF0ZS1waWNrZXJcbiAgICB2LW1vZGVsPVwidmFsdWVcIlxuICAgIHR5cGU9XCJkYXRlXCJcbiAgICBwbGFjZWhvbGRlcj1cIlBpY2sgYSBkYXlcIlxuICAgIDpkaXNhYmxlZC1kYXRlPVwiZGlzYWJsZWREYXRlXCJcbiAgLz5cbiAgPHA+e3sgdmFsdWU/LnRvTG9jYWxlU3RyaW5nKCkgfX08L3A+XG5cbiAgPGVsLWRhdGUtcGlja2VyXG4gICAgdi1tb2RlbD1cInZhbHVlMVwiXG4gICAgdHlwZT1cIm1vbnRoXCJcbiAgICBwbGFjZWhvbGRlcj1cIlBpY2sgYSBkYXlcIlxuICAgIDpkaXNhYmxlZC1kYXRlPVwiZGlzYWJsZWREYXRlXCJcbiAgLz5cbiAgPHA+e3sgdmFsdWUxPy50b0xvY2FsZVN0cmluZygpIH19PC9wPlxuXG4gIDxlbC1kYXRlLXBpY2tlclxuICAgIHYtbW9kZWw9XCJ2YWx1ZTJcIlxuICAgIHR5cGU9XCJtb250aHNcIlxuICAgIHBsYWNlaG9sZGVyPVwiUGljayBhIGRheVwiXG4gICAgOmRpc2FibGVkLWRhdGU9XCJkaXNhYmxlZERhdGVcIlxuICAvPlxuICA8cD57eyB2YWx1ZTI/LnRvTG9jYWxlU3RyaW5nKCkgfX08L3A+XG48L3RlbXBsYXRlPlxuIiwiZWxlbWVudC1wbHVzLmpzIjoiaW1wb3J0IHsgZ2V0Q3VycmVudEluc3RhbmNlIH0gZnJvbSAndnVlJ1xuaW1wb3J0IEVsZW1lbnRQbHVzIGZyb20gJ2VsZW1lbnQtcGx1cydcblxubGV0IGluc3RhbGxlZCA9IGZhbHNlXG5hd2FpdCBsb2FkU3R5bGUoKVxuXG5leHBvcnQgZnVuY3Rpb24gc2V0dXBFbGVtZW50UGx1cygpIHtcbiAgaWYgKGluc3RhbGxlZCkgcmV0dXJuXG4gIGNvbnN0IGluc3RhbmNlID0gZ2V0Q3VycmVudEluc3RhbmNlKClcbiAgaW5zdGFuY2UuYXBwQ29udGV4dC5hcHAudXNlKEVsZW1lbnRQbHVzKVxuICBpbnN0YWxsZWQgPSB0cnVlXG59XG5cbmV4cG9ydCBmdW5jdGlvbiBsb2FkU3R5bGUoKSB7XG4gIGNvbnN0IHN0eWxlcyA9IFsnaHR0cHM6Ly9wcmV2aWV3LTE1ODQ4LWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguY3NzJywgJ2h0dHBzOi8vcHJldmlldy0xNTg0OC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmNzcyddLm1hcCgoc3R5bGUpID0+IHtcbiAgICByZXR1cm4gbmV3IFByb21pc2UoKHJlc29sdmUsIHJlamVjdCkgPT4ge1xuICAgICAgY29uc3QgbGluayA9IGRvY3VtZW50LmNyZWF0ZUVsZW1lbnQoJ2xpbmsnKVxuICAgICAgbGluay5yZWwgPSAnc3R5bGVzaGVldCdcbiAgICAgIGxpbmsuaHJlZiA9IHN0eWxlXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2xvYWQnLCByZXNvbHZlKVxuICAgICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdlcnJvcicsIHJlamVjdClcbiAgICAgIGRvY3VtZW50LmJvZHkuYXBwZW5kKGxpbmspXG4gICAgfSlcbiAgfSlcbiAgcmV0dXJuIFByb21pc2UuYWxsU2V0dGxlZChzdHlsZXMpXG59XG4iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogXCJFU05leHRcIixcbiAgICBcImpzeFwiOiBcInByZXNlcnZlXCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJ0eXBlc1wiOiBbXCJlbGVtZW50LXBsdXMvZ2xvYmFsLmQudHNcIl0sXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiUGxheWdyb3VuZE1haW4udnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBBcHAgZnJvbSAnLi9BcHAudnVlJ1xuaW1wb3J0IHsgc2V0dXBFbGVtZW50UGx1cyB9IGZyb20gJy4vZWxlbWVudC1wbHVzLmpzJ1xuc2V0dXBFbGVtZW50UGx1cygpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8QXBwIC8+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge1xuICAgIFwidnVlXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3J1bnRpbWUtZG9tQGxhdGVzdC9kaXN0L3J1bnRpbWUtZG9tLmVzbS1icm93c2VyLmpzXCIsXG4gICAgXCJAdnVlL3NoYXJlZFwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9zaGFyZWRAbGF0ZXN0L2Rpc3Qvc2hhcmVkLmVzbS1idW5kbGVyLmpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctMTU4NDgtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJ1bnN1cHBvcnRlZFwiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsIl9vIjp7InNob3dIaWRkZW4iOnRydWUsInN0eWxlU291cmNlIjoiaHR0cHM6Ly9wcmV2aWV3LTE1ODQ4LWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguY3NzIn19)

Comment (xingyixiang) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:458]: type="months" 存在问题

Comment (xingyixiang) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:458]: 示例中: type=months的时切换到选择年份面板选中具体年份不会更新选择的值 只有选择了月份才会更新

之前的问题是
1、点击日期选择顶部入口【月份】【年份】
2、选择可选月份或年份，默认会显示月份第一天或第年份一个月第一日（此日期可能为禁止日期）

Comment (tolking) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:458]: type=months 时直接选择月份依然会选中 2024/8/1 00:00:00。这和 type=month 的行为不一致。我感觉这也是 disabledDate 相关的问题

Comment (xingyixiang) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue]: done

File: packages/components/date-picker/src/date-picker-com/panel-date-pick.vue [hunk 3]
```diff
@@ -452,12 +451,21 @@ const handleMonthPick = async (
...
+      innerDate.value.year(),
+      month as number
+    )
     emit(innerDate.value, false)
   } else if (selectionMode.value === 'months') {
-    emit(month as MonthsPickerEmits, keepOpen ?? true)
+    const validMonth = (month as MonthsPickerEmits).map((v) =>
+      getValidDateOfMonth(v.year(), v.month())
+    )
+    emit(validMonth, keepOpen ?? true)
```
Comment (tolking) [inline packages/components/date-picker/src/date-picker-com/panel-date-pick.vue:471]: 这样每次选择都会触发一次循环遍历，有没有办法直接在 basic-month-table.vue 中对新增加的值进行判断。这样应该可能更高效

Comment (xingyixiang) [reply packages/components/date-picker/src/date-picker-com/panel-date-pick.vue]: done

File: packages/components/date-picker/src/date-picker-com/basic-month-table.vue [hunk 4]
```diff
@@ -230,11 +224,15 @@ const handleMonthTableClick = (event: MouseEvent | KeyboardEvent) => {
...
+      month,
+      lang.value,
+      props.disabledDate
+    )
     const newValue = hasClass(target, 'current')
       ? castArray(props.parsedValue).filter(
-          (d) => Number(d) !== Number(newMonth)
+          (d) => d?.month() !== newMonth.month()
         )
       : castArray(props.parsedValue).concat([dayjs(newMonth)])
```
Comment (tolking) [inline packages/components/date-picker/src/date-picker-com/basic-month-table.vue:237]: Are `dayjs(newMonth)` and `newMonth` the same?

Comment (xingyixiang) [reply packages/components/date-picker/src/date-picker-com/basic-month-table.vue:237]: e090a0038e1213877e30021c33f993f
4e1122a4c1d20f07afc10f9a6c39f48
yes

**Review by tolking — APPROVED** [review]

**Comment by tolking** [comment]
感谢你的贡献

**Pull request merged**